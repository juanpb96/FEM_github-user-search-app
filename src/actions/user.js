import { types } from "../types/types";

export const setUserData = (userData) => ({
    type: types.setUserData,
    payload: userData
});