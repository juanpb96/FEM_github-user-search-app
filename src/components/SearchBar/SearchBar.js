import { useEffect, useRef, useState } from "react";
import { useDispatch } from "react-redux";
import { setUserData } from "../../actions/user";
import { getUser } from "../../helpers/getUser";


export const SearchBar = () => {

    const dispatch = useDispatch();

    const [username, setUsername] = useState('');
    const [errorState, setErrorState] = useState('');

    const errorRef = useRef(null);

    useEffect(() => {
        if (errorRef && errorRef.current) {
            if (errorState === 'active') {
                errorRef.current.focus();
            }

            if (errorState === 'closing') {
                errorRef.current.addEventListener('animationend', () => {
                    setErrorState('');
                }, { once: true });
            }
        }
    }, [errorRef, errorState]);
    
    const handleInputChange = ({target}) => {
        setUsername(target.value);
    };

    const handleSubmit = async(e) => {
        e.preventDefault();
        const userData = await getUser( username.trim() );

        if ( !userData ) {
            setErrorState('active');
            return;
        }

        dispatch( setUserData( userData ) );
        setErrorState('closing');
    };

    return (
        <form 
            className='[ search-bar ] [ flex-flow ] [ bg-secondary card-border mb-16 ]'
            onSubmit={ handleSubmit }
        >
            <img 
                className='icon'
                src={`${ process.env.PUBLIC_URL }/assets/icon-search.svg`}
                alt='' 
            />
            <input 
                type='text'
                name='username'
                value={ username }
                onChange={ handleInputChange }
                placeholder='Search GitHub usernameâ€¦'
                className='[ search-bar__input ] [ text-color-secondary fw-400 ]'
            />
            <p 
                className='[ search-bar__error ] [ text-error-color fs-small fw-700 ml-auto ]'
                datatype={ errorState ? errorState : undefined }
                ref={ errorRef }
                tabIndex='0'
            >No results</p>
            <button 
                type='submit'
                className='[ btn ] [ text-color-white fw-700 br-10 ml-auto ]'
                datatype='blue'
            >Search</button>
        </form>
    );
};
