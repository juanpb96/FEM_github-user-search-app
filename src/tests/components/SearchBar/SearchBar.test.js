import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import { SearchBar } from '../../../components/SearchBar/SearchBar';
import { Provider } from 'react-redux';
import { setUserData } from '../../../actions/user';
import { act, waitFor } from '@testing-library/react';

jest.mock('../../../actions/user', () => ({
    setUserData: jest.fn()
}));

const mockStore = configureStore([]);

const store = mockStore();
store.dispatch = jest.fn();

describe('Test <SearchBar />', () => { 
    const wrapper = mount(
        <Provider store={store}>
            <SearchBar />
        </Provider>
    );

    afterEach(() => {
        global.fetch = undefined;
    })

    test('should render correctly', () => {     
        expect(wrapper).toMatchSnapshot();
    });

    test('should call setUserData action and get the response correctly', async() => { 
        const mockResponse = {
            name: 'octocat',
            bio: null,
            company: '@github'
        };

        global.fetch = jest.fn(() =>
            Promise.resolve({
                json: () => Promise.resolve(mockResponse),
            })
        );

        await wrapper.find('form').prop('onSubmit')({ preventDefault(){} });

        expect(setUserData).toHaveBeenCalledTimes(1);
        expect(setUserData).toHaveBeenCalledWith(mockResponse);
    });

    test('should not call setUserData if the user does not exist and display an error', async() => { 
        const mockResponse = {
            message: 'Not Found'
        };
        
        global.fetch = jest.fn(() =>
            Promise.resolve({
                json: () => Promise.resolve(mockResponse),
            })
        );

        await act(async() => wrapper.find('form').prop('onSubmit')({ preventDefault(){} }));

        expect(setUserData).not.toHaveBeenCalled();
        expect(
            wrapper
            .find('.search-bar__error')
            .html()
            .includes('datatype="active"')
        ).toBe(true);
    });
});