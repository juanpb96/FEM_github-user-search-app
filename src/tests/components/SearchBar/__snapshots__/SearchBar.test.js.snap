// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <SearchBar /> should render correctly 1`] = `
<form
  className="[ search-bar ] [ flex-flow ] [ bg-secondary card-border mb-16 ]"
  onSubmit={[Function]}
>
  <img
    alt=""
    className="icon"
    height="20"
    src="/assets/icon-search.svg"
    width="20"
  />
  <input
    className="[ search-bar__input ] [ text-color-secondary fw-400 ]"
    name="username"
    onChange={[Function]}
    placeholder="Search GitHub usernameâ€¦"
    type="text"
    value=""
  />
  <p
    className="[ search-bar__error ] [ text-error-color fs-small fw-700 ml-auto ]"
    tabIndex="0"
  >
    No results
  </p>
  <button
    className="[ btn ] [ text-color-white fw-700 br-10 ml-auto ]"
    datatype="blue"
    type="submit"
  >
    Search
  </button>
</form>
`;
